test/Permissions.ts

```diff
@@ -0,0 +1,311 @@
+import test from 'fresh-tape'
+import { randomBytes } from 'crypto'
+import { Permissions } from '../src/Permissions'
+import { Operation, Request, ID, Response, ResponseType } from '../src/member'
+import HLC, { Timestamp } from '@consento/hlc'
+
+const memberA = 'a'
+const memberB = 'b'
+const memberC = 'c'
+const memberD = 'd'
+const memberE = 'e'
+const hlc = new HLC()
+
+function request (r: Partial<Request> & { operation: Operation, who: ID }): Request {
+  return {
+    type: 'request',
+    id: randomBytes(5).toString(),
+    from: memberA,
+    timestamp: hlc.now(),
+    ...r
+  }
+}
+
+function response (r: Partial<Response> & { response: ResponseType }): Response {
+  return {
+    type: 'response',
+    from: memberA,
+    timestamp: hlc.now(),
+    id: randomBytes(5).toString(),
+    ...r
+  }
+}
+
+test('First initialization', t => {
+  const p = new Permissions()
+  const req = request({ operation: 'add', who: memberA, from: memberA })
+  t.equals(p.add(req), req)
+  t.deepEquals(p.members.byState.added, new Set(memberA))
+  t.end()
+})
+
+test('Cant initialize with a remove request', t => {
+  const p = new Permissions()
+  t.throws(() => p.add(request({ operation: 'remove', who: memberA, from: memberA })), /First request needs to be an add request./)
```

Comment:
Is there a way we could use a predefined list of error messages or codes to make this easier?

Reply:
Totally! Should be on the task list for the whole system before release.

Reply:
SGTM! Mind adding a TODO to the main branch for that?